File: matlab/tests/unit/test_INLsine.m
%% test_INLsine.m - Unit test for INLsine function
% Tests the INLsine function to compute INL/DNL from sinewave data
%
% Output structure:
%   test_output/<data_set_name>/test_INLsine/
%       inl_dnl_matlab.csv      - code, INL, DNL
%       metrics_matlab.csv      - max_INL, min_INL, max_DNL, min_DNL
%       INLsine_matlab.png

close all; clc; clear;

%% Configuration
inputDir = "dataset";
outputDir = "test_output";

% Test datasets - leave empty to auto-search
filesList = {};
filesList = autoSearchFiles(filesList, inputDir, 'sinewave_*.csv');

if ~isfolder(outputDir)
    mkdir(outputDir);
end

%% Test Loop
fprintf('=== test_INLsine.m ===\n');
fprintf('[Testing] %d datasets...\n\n', length(filesList));

for k = 1:length(filesList)
    currentFilename = filesList{k};
    dataFilePath = fullfile(inputDir, currentFilename);

    if ~isfile(dataFilePath)
        fprintf('[%d/%d] %s - NOT FOUND, skipping\n\n', k, length(filesList), currentFilename);
        continue;
    end
    fprintf('[%d/%d] [Processing] %s\n', k, length(filesList), currentFilename);

    read_data = readmatrix(dataFilePath);

    % Extract dataset name
    [~, datasetName, ~] = fileparts(currentFilename);
    titleString = replace(datasetName, '_', '\_');

    % Create output subfolder
    subFolder = fullfile(outputDir, datasetName, 'test_INLsine');
    if ~isfolder(subFolder)
        mkdir(subFolder);
    end

    %% Run INLsine
    resolution = 12;
    [INL, DNL, code] = INLsine(read_data*2^resolution, 0.01);

    %% Plot
    figure('Visible', 'off');
    subplot(2,1,1);
    bar(code, DNL);
    grid on;
    xlabel('Code');
    ylabel('DNL (LSB)');
    title(['DNL: ', titleString]);
    xlim([0, 2^resolution])

    subplot(2,1,2);
    plot(code, INL, 'b-', 'LineWidth', 1);
    grid on;
    xlabel('Code');
    ylabel('INL (LSB)');
    title(['INL: ', titleString]);
    xlim([0, 2^resolution])


    % Save plot
    plotPath = fullfile(subFolder, 'INLsine_matlab.png');
    saveas(gcf, plotPath);
    fprintf('  [Saved] %s\n', plotPath);
    close(gcf);

    % Save INL/DNL data to CSV
    inlDnlTable = table(code', DNL', INL', ...
        'VariableNames', {'code', 'DNL', 'INL'});
    inlDnlPath = fullfile(subFolder, 'inl_dnl_matlab.csv');
    writetable(inlDnlTable, inlDnlPath);
    fprintf('  [Saved] %s\n', inlDnlPath);
    % Save metrics
    max_DNL = max(DNL);
    min_DNL = min(DNL);
    max_INL = max(INL);
    min_INL = min(INL);
    metricsTable = table(max_DNL, min_DNL, max_INL, min_INL, ...
        'VariableNames', {'max_DNL', 'min_DNL', 'max_INL', 'min_INL'});
    metricsPath = fullfile(subFolder, 'metrics_matlab.csv');
    writetable(metricsTable, metricsPath);
    fprintf('  [Saved] %s\n', metricsPath);
    fprintf('  [Results] DNL: [%.2f, %.2f], INL: [%.2f, %.2f] LSB\n\n', ...
        min_DNL, max_DNL, min_INL, max_INL);
end

fprintf('[test_INLsine COMPLETE]\n');

File: python/tests/unit/test_inl_sine.py
import numpy as np
import matplotlib.pyplot as plt

from adctoolbox.aout import inl_sine
from tests._utils import save_fig
from tests.unit._runner import run_unit_test_batch

plt.rcParams['font.size'] = 14
plt.rcParams['axes.grid'] = True

# Resolution list to scan
RESOLUTION_LIST = [12]

def _process_inl_sine(raw_data, sub_folder, dataset_name, figures_folder, test_name):
    """
    Callback function to process a single file:
    1. Scale data by resolution
    2. Calculate INL/DNL
    3. Generate and save plot
    """
    for Resolution in RESOLUTION_LIST:
        # 1. Scale data by 2^Resolution
        scaled_data = raw_data * (2 ** Resolution)
        expected_max = 2 ** Resolution

        # 2. Calculate INL/DNL
        INL, DNL, code = inl_sine(scaled_data)

        # Calculate ranges for plot titles
        max_inl = np.max(INL)
        min_inl = np.min(INL)
        max_dnl = np.max(DNL)
        min_dnl = np.min(DNL)

        # 3. Generate Plot
        fig = plt.figure(figsize=(10, 8))

        # Top subplot: INL
        plt.subplot(2, 1, 1)
        plt.scatter(code, INL, s=8, alpha=0.6)
        plt.xlabel('Code')
        plt.ylabel('INL (LSB)')
        plt.grid(True)
        plt.title(f'INL = [{min_inl:.2f}, {max_inl:+.2f}] LSB')
        ylim_min = min(min_inl, -1)
        ylim_max = max(max_inl, 1)
        plt.ylim([ylim_min, ylim_max])
        plt.xlim([0, expected_max])

        # Bottom subplot: DNL
        plt.subplot(2, 1, 2)
        plt.scatter(code, DNL, s=8, alpha=0.6)
        plt.xlabel('Code')
        plt.ylabel('DNL (LSB)')
        plt.grid(True)
        plt.title(f'DNL = [{min_dnl:.2f}, {max_dnl:.2f}] LSB')
        ylim_min = min(min_dnl, -1)
        ylim_max = max(max_dnl, 1)
        plt.ylim([ylim_min, ylim_max])
        plt.xlim([0, expected_max])

        plt.tight_layout()

        # 4. Save Figure
        figure_name = f"{dataset_name}_{test_name}_python.png"
        save_fig(figures_folder, figure_name, dpi=150)
        plt.close(fig)

def test_inl_sine(project_root):
    """
    Batch runner for inl_sine (Single Channel Version).
    """
    run_unit_test_batch(
        project_root=project_root,
        input_subpath="dataset/aout/sinewave",
        test_module_name="test_inl_sine",
        file_pattern="sinewave_*.csv",
        output_subpath="test_output",
        process_callback=_process_inl_sine
    )
